Lock File not properly generated - framework ordering should not impact the restore with lockedmode

## Details about Problem

**NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe):**
VS UI, MSBuild CLI, Nuget.exe

**NuGet version (x.x.x.xxx):**
- nuget version  5.3.0.6260
- MSBuild  version - 16.3.0+0f4c62fea for .NET Framework
   - Nuget CommonExtension installed version - Nuget.Commands.dll - 5.3.0.6251

**VS version (if appropriate):**
 [Information](https://gist.github.com/carpnick/a275bf0be54960cfa8892eeee9916e0c#file-vs-versioning) 

**OS version (i.e. win10 v1607 (14393.321)):**
10.0.17763 Build 17763 -  Windows Server 2019 Datacenter

**Worked before? If so, with which NuGet version:**
MSbuild based restore - 15.9.21+g9802d43bc3
  - Project dependend on - NuGet.Build.Tasks.Pack 4.9.2

## Detailed repro steps so we can see the same problem

Run all these steps from within Visual Studio 2019 (See version info)
1. Create a new .NET solution with a class library (full framework).
2. Add a nuget reference to Newtonsoft.json (any will do)
3. Edit csproj file and add `<RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>`
4. Reload project.  Build.
5.  Validate the packages.lock.json is generated.  See [example](https://gist.github.com/carpnick/a275bf0be54960cfa8892eeee9916e0c#file-packages-lock-json).
6. Edit CSPROJ again, this time add the following line:
` <RestoreLockedMode>true</RestoreLockedMode>`
7. Reload the project, and attempt to rebuild, 2 times.

Error Received:
`Error	NU1004	The packages lock file is inconsistent with the project dependencies so restore can't be run in locked mode. Disable the RestoreLockedMode MSBuild property or pass an explicit --force-evaluate option to run restore to update the lock file.			1		
`

MSBuild CLI to reproduce: `& 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\MSBuild.exe' /t:"clean;Restore" /p:Configuration="Release" /p:Platform="Any CPU" /p:RestoreLockedMode="true" /p:RestoreNoCache="true"`

NOTE: We did not stumble on this issue like this, we stumbled on this setting RestoreLockedMode in our CI process to true.  But the above steps produce a reliable easy way to reproduce.


## Other suggested things

Other notes:

- When you allow MSBuild or VS to ForceEvaluate - it keeps generating incorrect package lock files.
- nuget.exe works as expected  - CLI `C:\nuget.exe restore -lockedmode -NoCache`
- **What is broken** - Copied Nuget.Clients Source privately from this [tag](https://github.com/NuGet/NuGet.Client/tree/5.3.0.6251).
    - At [this location](https://github.com/NuGet/NuGet.Client/blob/5.3.0.6251/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs#L105-L108) I added [this code](https://gist.github.com/carpnick/a275bf0be54960cfa8892eeee9916e0c#file-privatepatch) to my private copy.  When I run with my private version with Visual Studio or MSBuild CLI, I get the [following](https://gist.github.com/carpnick/a275bf0be54960cfa8892eeee9916e0c#file-msbuild-custom-output).  Notice the order is different.  When I run with my private version of nuget.exe I get [this](https://gist.github.com/carpnick/a275bf0be54960cfa8892eeee9916e0c#file-custom-nuget-exe).

### Workaround available
 - Modify lock file after it is generated and modify the order of the .NETFrameworks. (This would be very painful from a developer point of view, but does offer a workaround).
 - Another option is to not use MSbuild /t:restore and instead use nuget.exe

### Sample Project

Very helpful if you can zip a project and paste into this issue!
[ExampleSolution.zip](https://github.com/NuGet/Home/files/3672626/ExampleSolution.zip)


PS:
This was very helpful in debugging down to where the issue was: https://github.com/NuGet/Home/issues/8493#issuecomment-526375416

Thanks @nkolev92

Yes @nkolev92  which is why I am so confused.  

I was testing with 16.3.0+0f4c62fea of MSBuild which has the DLL in the CommonExtensions folder with a version of 5.3.0.6251....

However, when I test the same exact solution with nuget.exe  which is version 5.3.0.6260, it is working...  I know not the exact same build but close.


Nuget.exe output:
```
c:\temp\Eliza-Eliza.Sample.VS2017.MSI>C:\nuget.exe restore -lockedmode -NoCache
MSBuild auto-detection: using msbuild version '16.3.0.46305' from 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\bin'.
Restoring packages for c:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib.unittests\Utilitylib.unittests.csproj...
Restoring packages for c:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib\Eliza.Sample.VS2017.MSI.Utilitylib.csproj...
Committing restore...
Committing restore...
Writing assets file to disk. Path: c:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib.unittests\obj\project.assets.json
Assets file has not changed. Skipping assets file writing. Path: c:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib\obj\project.assets.json
Restore completed in 726.6 ms for c:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib\Eliza.Sample.VS2017.MSI.Utilitylib.csproj.
Restore completed in 726.54 ms for c:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib.unittests\Utilitylib.unittests.csproj.

NuGet Config files used:
    C:\Users\Administrator\AppData\Roaming\NuGet\NuGet.Config
    C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

Feeds used:
    C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
    https://api.nuget.org/v3/index.json
```


MSBUILD:
```
PS C:\temp\Eliza-Eliza.Sample.VS2017.MSI> & 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\MSBuild.exe' /t:"Restore" /p:Configuration="Release" /p:Platform="Any CPU" /p:RestoreLockedMode="true" /p:RestoreNoCache="true"
Microsoft (R) Build Engine version 16.3.0+0f4c62fea for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
Build started 9/30/2019 9:18:45 PM.
Project "C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Eliza.Sample.VS2017.MSI.solution.sln" on node 1 (Restore target(s)).
ValidateSolutionConfiguration:
  Building solution configuration "Release|Any CPU".
ValidateProjects:
  The project "test1" is not selected for building in solution configuration "Release|Any CPU".
C:\temp\Eliza-Eliza.Sample.VS2017.MSI\SetupProject1\SetupProject1.wixproj : warning NU1503: Skipping restore for project 'C:\temp\Eliza-Eliza.Sample.VS2017.MSI\SetupProject1\SetupProject1.wixproj'. The project file may be invalid or mis
sing targets required for restore. [C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Eliza.Sample.VS2017.MSI.solution.sln]
Restore:
  Restoring packages for C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib.unittests\Utilitylib.unittests.csproj...
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets(123,5): error NU1004: The packages lock file is inconsistent with the project dependencies so restore can't be run
 in locked mode. Disable the RestoreLockedMode MSBuild property or pass an explicit --force-evaluate option to run restore to update the lock file. [C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Eliza.Sample.VS2017.MSI.solution.sln]
  Committing restore...
  Assets file has not changed. Skipping assets file writing. Path: C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib\obj\project.assets.json
  Writing packages lock file at disk. Path: C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib\packages.lock.json
  Restore failed in 90.33 ms for C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib\Eliza.Sample.VS2017.MSI.Utilitylib.csproj.
  Committing restore...
  Writing assets file to disk. Path: C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib.unittests\obj\project.assets.json
  Restore completed in 476.04 ms for C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Utilitylib.unittests\Utilitylib.unittests.csproj.

  NuGet Config files used:
      C:\Users\Administrator\AppData\Roaming\NuGet\NuGet.Config
      C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

  Feeds used:
      https://api.nuget.org/v3/index.json
      C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
Done Building Project "C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Eliza.Sample.VS2017.MSI.solution.sln" (Restore target(s)) -- FAILED.


Build FAILED.

"C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Eliza.Sample.VS2017.MSI.solution.sln" (Restore target) (1) ->
(_FilterRestoreGraphProjectInputItems target) ->
  C:\temp\Eliza-Eliza.Sample.VS2017.MSI\SetupProject1\SetupProject1.wixproj : warning NU1503: Skipping restore for project 'C:\temp\Eliza-Eliza.Sample.VS2017.MSI\SetupProject1\SetupProject1.wixproj'. The project file may be invalid or m
issing targets required for restore. [C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Eliza.Sample.VS2017.MSI.solution.sln]


"C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Eliza.Sample.VS2017.MSI.solution.sln" (Restore target) (1) ->
(Restore target) ->
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets(123,5): error NU1004: The packages lock file is inconsistent with the project dependencies so restore can't be r
un in locked mode. Disable the RestoreLockedMode MSBuild property or pass an explicit --force-evaluate option to run restore to update the lock file. [C:\temp\Eliza-Eliza.Sample.VS2017.MSI\Eliza.Sample.VS2017.MSI.solution.sln]

    1 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.28
```

--- a/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs
+++ b/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs
@@ -105,8 +105,6 @@ namespace NuGet.ProjectModel
             var projectRuntimesKeys = project.RuntimeGraph.Runtimes.Select(r => r.Key).Where(k => k != null);
             var lockFileRuntimes = nuGetLockFile.Targets.Select(t => t.RuntimeIdentifier).Where(r => r != null).Distinct();
 
-
-            //BUG: 8661 - Making it order both lists so we dont have this bug again with SequenceEquals
             if (!projectRuntimesKeys.OrderedEquals(
                             lockFileRuntimes,
                             x => x,

--- a/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs
+++ b/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs
@@ -111,10 +111,11 @@ namespace NuGet.ProjectModel
             //Based on the above code, we want to order the reading side.
             //Making it order both lists so we dont have this bug again with SequenceEquals
             if (!projectRuntimesKeys.OrderedEquals(
-                                                    lockFileRuntimes,
-                                                    x => x,
-                                                    Comparer<string>.Default,
-                                                    EqualityComparer<string>.Default))
+                            lockFileRuntimes,
+                            x => x,
+                            StringComparer.InvariantCultureIgnoreCase,
+                            StringComparer.InvariantCultureIgnoreCase))
+
             {
                 return false;
             }

--- a/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs
+++ b/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs
@@ -105,17 +105,13 @@ namespace NuGet.ProjectModel
             var projectRuntimesKeys = project.RuntimeGraph.Runtimes.Select(r => r.Key).Where(k => k != null);
             var lockFileRuntimes = nuGetLockFile.Targets.Select(t => t.RuntimeIdentifier).Where(r => r != null).Distinct();
 
-            ;
 
-            //JsonRuntimeFormat.WriteRuntimeGraph is doing a sort based on runtimeIdentifier
-            //Based on the above code, we want to order the reading side.
-            //Making it order both lists so we dont have this bug again with SequenceEquals
+            //BUG: 8661 - Making it order both lists so we dont have this bug again with SequenceEquals
             if (!projectRuntimesKeys.OrderedEquals(
                             lockFileRuntimes,
                             x => x,
                             StringComparer.InvariantCultureIgnoreCase,
                             StringComparer.InvariantCultureIgnoreCase))
-
             {
                 return false;
             }

--- a/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs
+++ b/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs
@@ -105,7 +105,16 @@ namespace NuGet.ProjectModel
             var projectRuntimesKeys = project.RuntimeGraph.Runtimes.Select(r => r.Key).Where(k => k != null);
             var lockFileRuntimes = nuGetLockFile.Targets.Select(t => t.RuntimeIdentifier).Where(r => r != null).Distinct();
 
-            if (!projectRuntimesKeys.SequenceEqual(lockFileRuntimes))
+            ;
+
+            //JsonRuntimeFormat.WriteRuntimeGraph is doing a sort based on runtimeIdentifier
+            //Based on the above code, we want to order the reading side.
+            //Making it order both lists so we dont have this bug again with SequenceEquals
+            if (!projectRuntimesKeys.OrderedEquals(
+                                                    lockFileRuntimes,
+                                                    x => x,
+                                                    Comparer<string>.Default,
+                                                    EqualityComparer<string>.Default))
             {
                 return false;
             }

--- a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
+++ b/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
@@ -9021,6 +9021,7 @@ namespace NuGet.CommandLine.Test
                 var originalTargets = packagesLockFile.Targets.Where(t => t.RuntimeIdentifier != null).Select(t => t.RuntimeIdentifier).ToList();
                 runtimeidentifiers.ShouldBeEquivalentTo(originalTargets);
 
+                //Nuget.exe test so reordering to make it not match.  It should still restore correctly
                 packagesLockFile.Targets = packagesLockFile.Targets.
                     OrderByDescending(t => t.RuntimeIdentifier==null).
                     ThenByDescending( i => i.RuntimeIdentifier).ToList();

--- a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
+++ b/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
@@ -8958,6 +8958,77 @@ namespace NuGet.CommandLine.Test
             }
         }
 
+
+
+        [Fact]
+        public async void RestoreNetCore_PackagesLockFile_WithReorderedRuntimesInLockFile_PassRestore() { 
+            // A project with RestoreLockedMode should pass restore if the runtimes in the lock file have been reordered
+            using (var pathContext = new SimpleTestPathContext())
+            {
+
+                // Set up solution, project, and packages
+                var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+                var projFramework = Frameworks.FrameworkConstants.CommonFrameworks.Net461.GetShortFolderName();
+
+                var projectA = SimpleTestProjectContext.CreateNETCore(
+                   "a",
+                   pathContext.SolutionRoot,
+                   NuGetFramework.Parse(projFramework));
+
+                projectA.Properties.Add("RestorePackagesWithLockFile", "true");
+
+                // Set up the package and source
+                var packageX = new SimpleTestPackageContext()
+                {
+                    Id = "x",
+                    Version = "1.0.0"
+                };
+                packageX.Files.Clear();
+                packageX.AddFile("lib/netcoreapp2.0/x.dll");
+                packageX.AddFile("ref/netcoreapp2.0/x.dll");
+                packageX.AddFile("lib/net461/x.dll");
+                packageX.AddFile("ref/net461/x.dll");
+
+                await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                    pathContext.PackageSource,
+                    PackageSaveMode.Defaultv3,
+                    packageX);
+
+                projectA.AddPackageToAllFrameworks(packageX);
+
+
+                solution.Projects.Add(projectA);
+                solution.Create(pathContext.SolutionRoot);
+
+                // Act
+                var r = Util.RestoreSolution(pathContext);
+
+
+                Assert.True(File.Exists(projectA.NuGetLockFileOutputPath));
+                var packagesLockFile = PackagesLockFileFormat.Read(projectA.NuGetLockFileOutputPath);
+
+                //Modify the list of target/runtimes so they are reordered in the lock file.
+                var oneOfTargets = packagesLockFile.Targets[1];
+                packagesLockFile.Targets.RemoveAt(1);
+                packagesLockFile.Targets.Insert(packagesLockFile.Targets.Count - 1, oneOfTargets);
+
+                PackagesLockFileFormat.Write(projectA.NuGetLockFileOutputPath, packagesLockFile);
+                projectA.Properties.Add("RestoreLockedMode", "true");
+                projectA.Save();
+
+
+                //Run the restore and it should still properly restore.
+                var r2 = Util.RestoreSolution(pathContext);
+
+                // Assert
+                r.Success.Should().BeTrue();
+                
+            }
+        }
+
+        
+
+
         private static byte[] GetTestUtilityResource(string name)
         {
             return ResourceTestUtility.GetResourceBytes(

--- a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
+++ b/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
@@ -8968,12 +8968,19 @@ namespace NuGet.CommandLine.Test
 
                 // Set up solution, project, and packages
                 var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
-                var projFramework = Frameworks.FrameworkConstants.CommonFrameworks.Net461.GetShortFolderName();
+               
+
+                var projFramework = Frameworks.FrameworkConstants.CommonFrameworks.NetCoreApp21.GetShortFolderName();
 
                 var projectA = SimpleTestProjectContext.CreateNETCore(
                    "a",
                    pathContext.SolutionRoot,
                    NuGetFramework.Parse(projFramework));
+
+                var runtimeidentifiers = new List<string>(){ "win7-x64", "win-x86", "win", "z", "a" };
+                var ascending = runtimeidentifiers.OrderBy(i => i);
+                
+                projectA.Properties.Add("RuntimeIdentifiers", string.Join(";", ascending));
 
                 projectA.Properties.Add("RestorePackagesWithLockFile", "true");
 
@@ -9007,11 +9014,22 @@ namespace NuGet.CommandLine.Test
                 Assert.True(File.Exists(projectA.NuGetLockFileOutputPath));
                 var packagesLockFile = PackagesLockFileFormat.Read(projectA.NuGetLockFileOutputPath);
 
-                //Modify the list of target/runtimes so they are reordered in the lock file.
-                var oneOfTargets = packagesLockFile.Targets[1];
-                packagesLockFile.Targets.RemoveAt(1);
-                packagesLockFile.Targets.Insert(packagesLockFile.Targets.Count - 1, oneOfTargets);
+                //Modify the list of target/runtimes so they are reordered in the lock file
+                //Verify the passed in RIDs are within the lock file
+                //Verify the RIDS are not the same after reordering.
+                //Lock file is not ordered based on input RIDs so Validating the reorder here.
+                var originalTargets = packagesLockFile.Targets.Where(t => t.RuntimeIdentifier != null).Select(t => t.RuntimeIdentifier).ToList();
+                runtimeidentifiers.ShouldBeEquivalentTo(originalTargets);
 
+                packagesLockFile.Targets = packagesLockFile.Targets.
+                    OrderByDescending(t => t.RuntimeIdentifier==null).
+                    ThenByDescending( i => i.RuntimeIdentifier).ToList();
+                var reorderedTargets = packagesLockFile.Targets.Where(t => t.RuntimeIdentifier != null).Select(t => t.RuntimeIdentifier).ToList();
+
+                //The orders are not equal.  Then resave the lock file and project.
+                //The null RID must be the first one otherwise this fails
+                Assert.False(originalTargets.SequenceEqual(reorderedTargets));                
+                Assert.True(packagesLockFile.Targets[0].RuntimeIdentifier == null);
                 PackagesLockFileFormat.Write(projectA.NuGetLockFileOutputPath, packagesLockFile);
                 projectA.Properties.Add("RestoreLockedMode", "true");
                 projectA.Save();

--- a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
+++ b/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
@@ -9008,6 +9008,7 @@ namespace NuGet.CommandLine.Test
                 var packagesLockFile = PackagesLockFileFormat.Read(projectA.NuGetLockFileOutputPath);
 
                 //Modify the list of target/runtimes so they are reordered in the lock file.
+                Assert.True(packagesLockFile.Targets.Count >=3); //Since we are removing the second one and adding to the end.
                 var oneOfTargets = packagesLockFile.Targets[1];
                 packagesLockFile.Targets.RemoveAt(1);
                 packagesLockFile.Targets.Insert(packagesLockFile.Targets.Count - 1, oneOfTargets);

--- a/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/Dotnet.Integration.Test.csproj
+++ b/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/Dotnet.Integration.Test.csproj
@@ -13,6 +13,7 @@
   </ItemGroup>
   
   <ItemGroup>
+    <ProjectReference Include="..\..\..\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj" />
     <ProjectReference Include="..\NuGet.XPlat.FuncTest\NuGet.XPlat.FuncTest.csproj" />
     <ProjectReference Include="$(TestUtilitiesDirectory)Test.Utility\Test.Utility.csproj" />
   </ItemGroup>
@@ -21,6 +22,6 @@
     <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
   </ItemGroup>
 
-  <Import Project="$(BuildCommonDirectory)common.targets"/>
+  <Import Project="$(BuildCommonDirectory)common.targets" />
   <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
 </Project>

--- a/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/Dotnet.Integration.Test.csproj
+++ b/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/Dotnet.Integration.Test.csproj
@@ -13,7 +13,6 @@
   </ItemGroup>
   
   <ItemGroup>
-    <ProjectReference Include="..\..\..\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj" />
     <ProjectReference Include="..\NuGet.XPlat.FuncTest\NuGet.XPlat.FuncTest.csproj" />
     <ProjectReference Include="$(TestUtilitiesDirectory)Test.Utility\Test.Utility.csproj" />
   </ItemGroup>

--- a/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs
+++ b/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs
@@ -9,6 +9,7 @@ using System.Xml.Linq;
 using FluentAssertions;
 using NuGet.Common;
 using NuGet.Frameworks;
+using NuGet.Packaging;
 using NuGet.Test.Utility;
 using Xunit;
 
@@ -284,7 +285,75 @@ EndGlobal";
             }
         }
 
-        private static byte[] GetResource(string name)
+        [PlatformFact(Platform.Windows)]
+        public async Task DotnetRestore_LockedMode_NewProjectOutOfBox()
+        {
+            using (var pathContext = new SimpleTestPathContext())
+            {
+
+                // Set up solution, and project
+                var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+
+
+                var projFramework = FrameworkConstants.CommonFrameworks.Net462.GetShortFolderName();
+
+                var projectA = SimpleTestProjectContext.CreateNETCore(
+                   "a",
+                   pathContext.SolutionRoot,
+                   NuGetFramework.Parse(projFramework));
+
+                var runtimeidentifiers = new List<string>() { "win7-x64", "win-x86", "win" };
+                projectA.Properties.Add("RuntimeIdentifiers", string.Join(";", runtimeidentifiers));
+                projectA.Properties.Add("RestorePackagesWithLockFile", "true");
+
+                //Setup packages and feed
+                var packageX = new SimpleTestPackageContext()
+                {
+                    Id = "x",
+                    Version = "1.0.0"
+                };
+                packageX.Files.Clear();
+                packageX.AddFile("lib/netcoreapp2.0/x.dll");
+                packageX.AddFile("ref/netcoreapp2.0/x.dll");
+                packageX.AddFile("lib/net461/x.dll");
+                packageX.AddFile("ref/net461/x.dll");
+
+                await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                    pathContext.PackageSource,
+                    PackageSaveMode.Defaultv3,
+                    packageX);
+
+
+                //add the packe to the project
+                projectA.AddPackageToAllFrameworks(packageX);
+                solution.Projects.Add(projectA);
+                solution.Create(pathContext.SolutionRoot);
+                solution.Save();
+                projectA.Save();
+
+
+                // Act
+                var args = $" --source \"{pathContext.PackageSource}\" ";
+                var projdir = Path.GetDirectoryName(projectA.ProjectPath);
+                var projfilename = Path.GetFileNameWithoutExtension(projectA.ProjectName);
+
+                _msbuildFixture.RestoreProject(projdir, projfilename, args);
+                Assert.True(File.Exists(projectA.NuGetLockFileOutputPath));
+
+                //Now set it to locked mode
+                projectA.Properties.Add("RestoreLockedMode", "true");
+                projectA.Save();
+
+                //Act
+                //Run the restore and it should still properly restore.
+                //Assert within RestoreProject piece
+                _msbuildFixture.RestoreProject(projdir, projfilename, args);
+                Assert.True(File.Exists(projectA.NuGetLockFileOutputPath));
+            }
+        }
+
+
+            private static byte[] GetResource(string name)
         {
             return ResourceTestUtility.GetResourceBytes(
                 $"Dotnet.Integration.Test.compiler.resources.{name}",

